src/Core/src/Hosting/ImageSources/ImageSourceServiceProvider.cs

```diff
@@ -14,16 +14,22 @@ sealed class ImageSourceServiceProvider : MauiFactory, IImageSourceServiceProvid
 		readonly ConcurrentDictionary<Type, Type> _imageSourceCache = new ConcurrentDictionary<Type, Type>();
 		readonly ConcurrentDictionary<Type, Type> _serviceCache = new ConcurrentDictionary<Type, Type>();
 
+		readonly ImageSourceToImageSourceServiceTypeMapping _imageSourceMapping;
+
 		public ImageSourceServiceProvider(IImageSourceServiceCollection collection, IServiceProvider hostServiceProvider)
 			: base(collection)
 		{
+			_imageSourceMapping = ImageSourceToImageSourceServiceTypeMapping.GetInstance(collection);
 			HostServiceProvider = hostServiceProvider;
 		}
 
 		public IServiceProvider HostServiceProvider { get; }
 
-		public IImageSourceService? GetImageSourceService(Type imageSource) =>
-			(IImageSourceService?)GetService(GetImageSourceServiceType(imageSource));
+		public IImageSourceService? GetImageSourceService(Type imageSource)
+		{
+			var imageSourceService = _serviceCache.GetOrAdd(imageSource, _imageSourceMapping.FindImageSourceServiceType);
+			return (IImageSourceService?)GetService(imageSourceService);
+		}
 
 		public Type GetImageSourceServiceType(Type imageSource) =>
```

Comment:
Should this method now be calling `_imageSourceMapping.FindImageSourceServiceType`?

Reply:
I'm still trying to figure out if it's better to keep the old behavior of this method but make it obsolete (but `GetImageSourceServiceType(imageSource)` might not necessarily match `GetImageSourceService(imageSource)?.GetType()`) or to change the way this method behaves and break someone's expectations. I'm going back and forth on this.